### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by Dherse** [typst/typst#712]
This issue is a tracking issue for many different new features and improvements to Typst.
They are listed below in no particular order, however some do depend on other issues that need to be addressed first.
Additionally, some of the issues are more complex than others, and will be split into multiple issues.
Finally, most issues will have separate RFCs that will be linked in this document as they are written.

## Layout
**Motivation:** Make the layout engine more capable and its behaviour more intuitive.

- [ ] Merge math and normal layout code
  - [ ] Improve baseline handling
  - [ ] Make normal layout containers behave sanely in math
- [ ] Rework block spacing, leading, and text edges
- [ ] Improve table layout
- [ ] Support out-of-flow content
  - [x] Footnotes
  - [ ] Floats
  - [ ] General placement with collision, e.g. wrapping figures
- [ ] Support richer page customization, like margins depending on page number
- [ ] Inside/outside alignment
- [ ] Support inserting custom content at breakpoints
- [ ] Chain multiple containers together to fill them with one flow of content
- [ ] Grid typesetting
- [ ] (To be completed)

## Content and Styling
**Motivation:** Enable HTML export and accessibility, support more powerful styling needs, fix various complex styling bugs and improve performance.

- [ ] Rework styling system
  - [ ] Split up layout phase into realization and layout
    - Realization deals with styling and building the elements
    - Layout computes the absolute positions (this part would be replaced for HTML export)
  - [ ] Move all layout rules into an appropriate set of distinct rules
  - [ ] Add acceleration structures for faster (and easier) lookup of styling information for a given element
- [ ] Rework the `Content` system with the following goal: (See #765)
  - [ ] Keep full track of styling
  - [ ] Keep full track of ancestry
  - [ ] Improve the internal representation (remove `Attr` if possible)
- [ ] Support custom components
  - [ ] Possibly expose capability system to users
- [ ] Rewrite some built-in components to use the custom component system instead (this makes the components more welcoming for non-rust developers and moves the "standard lib" into Typst as much as possible)
- [ ] Add additional selectors for more flexibility

## Development and Performance
**Motivation:** Improve the contribution experience and ensure good performance long-term.

- [ ] Instrumentation (See #757, #761)
  - [x] Adding instrumentation to every function that does large amounts of work
  - [ ] Adding logging to critical sections
- [ ] Benchmarks (See #756)
  - [ ] Measure overall compile times
  - [ ] Measure incremental compile times
  - [ ] Measure internals, such as: lexing, parsing, function calls, common binary operations, `Layout`, `show`, etc. as dedicated benchmarks
  - [ ] Keep track of changes in a publicly visible github site - the goal is to see regressions and improvements, systematically use the same hardware in order to make the test reliable and repeatable
  - [ ] Measure cache efficiency in comemo
- [ ] Documentation
  - [ ] Document all internal functions - this is key to making the codebase more welcoming to newcomers.
  - [ ] Create a global contribution documentation containing `rustdoc` of the internals (including a nightly build of the doc), additionally manually written documentation, published on the typst.app website
- [ ] General improvements (can be done in parallel)
  - [ ] Collect low-hanging optimization fruit, if any (find it with instrumentation)
  - [ ] Add more methods to make common actions simpler in rust code

## Further things
- [ ] Package manager - can be done in parallel with a lot of the work
- [ ] Support HTML export, first needs the new realization phase (see #721)
- [ ] Properly handle I18n

(Some of these should be expanded into separate categories.)

**Comment by astrale-sharp** [comment]
Great issue !
I'd argue to order the element in order of importance, my opinion for the importance is :
Primary : 
- Instrumenting typst using [tracing](https://docs.rs/tracing/latest/tracing/)
- Improved layout system
- Rework the Content system
- Fully document all internal functions 
Secondary :
- Package manager
- Add HTML export
- Create a global contribution documentation containing rustdoc of the internals
- Create custom components
- Adding benchmarks to typst using criterion - the defacto standard in the rust ecosystem

there is no order inside of primary secondary

**Comment by mattfbacon** [comment]
hm, in your prototype Content struct, why do you use NonNull and VTable instead of just `Box<dyn ElementTrait>`?

**Comment by Dherse** [comment]
> hm, in your prototype Content struct, why do you use NonNull and VTable instead of just `Box<dyn ElementTrait>`?

Actually, there is a difference, the VTable is at least needed for aditional marker traits, additionally since in my new design the trait has associated types, it cannot be made dyn. But you're right, Box<dyn Trait> is better and for the Element trait itself it should be like this with maybe an auto-impl trait called DynElement or something. I'm just bouncing ideas here and looking for feedback 😉

**Comment by SUPERCILEX** [comment]
For compile time benchmarks, criterion seems like a poor fit. I'd suggest [hyperfine](https://github.com/sharkdp/hyperfine) instead.

**Comment by alerque** [comment]
@SUPERCILEX Just to be clear, are you aware that this project uses "compile" for the act of using the software to render a document, not transform the source code into the executable? In other words "compile time" in the issue above is what programmers refer to as "run time" as opposed to "build time".

**Comment by SUPERCILEX** [comment]
Oh wait no, now I remember what I was thinking it meant: execute the typst binary to render a document. If that's what "compile" means in this context, then I stand by hyperfine since it better suited for that end-to-end benchmarking.

**Comment by astrale-sharp** [comment]
I'll start lightly experimenting with [hyperfine](https://github.com/sharkdp/hyperfine).

**Comment by ararunaufc** [comment]
I'd like to work on some of the layouting designing and changes, so I would love to be included in these discussions 😄 

This is the part of the code that I became most familiar with.

**Comment by imbillow** [comment]
Hi, is anyone implementing our typst's https://github.com/typst/typst/issues/765 feature?

**Comment by Dherse** [comment]
> Hi, is anyone implementing our typst's #765 feature?

Hey, currently I have done some work as part of a larger set of optimizations, however my initial work was very basic and only moved `Attr::Field` and `Attr::Value` into a dedicated field. One thing to keep in mind is that we'd like to keep `Content` a single allocation internally which is going to be rather difficult.

**Comment by Jmcjack** [comment]
Is there an official plan for citations and bibliography? It looks like hayagriva is not that active and is quite buggy from the IEEE styling point of view.

**Comment by sirinath** [comment]
Couple of features I would like is:
- ability to factor out styling elements  into an seperate file which can be easily changed
- grid typesetting
- local IDE for big projects with the ability organise facts and text snippets to incorporate in the final text

**Comment by mnrvwl** [comment]
Hey @Dherse, do you think https://github.com/typst/typst/issues/1139 as suitable to be included here?
I'm open to do any edits necessary to better match other RFCs. Cheers 🚀

**Comment by vrilcode** [comment]
> * grid typesetting

I totally agree. For professional typesetting, this is a "must have".

**Comment by sirinath** [comment]
@vrilcode  thanks for this: https://github.com/typst/typst/issues/1306

**Comment by laurmaedje** [comment]
Closing in favor of https://typst.app/docs/roadmap.
